File: static/algorithms/sorting/bubble_sort/c/bubble_sort.c
#include <stdio.h>  // for printf

const int MAX_NUMBERS = 10000;

void bubbleSort(int[], int);
int swapAsc(int *, int *);
void printIntArray(int, int []);

int main(int argc, char* argv[]) {
	const int size = 10;	// Number of items
	int numbers[size] = {2, 6, 3, 9, 8, 7, 1, 5, 4, 10};

    // Output unsorted numbers
	printf("Unsorted: \t");
	printIntArray(size, numbers);

    // Sort numbers
    bubbleSort(numbers, size);

    // Output sorted numbers
	printf("Sorted: \t");
	printIntArray(size, numbers);

    return 0;
}

/* Bubble Sort
 * Sorts elements by comparing adjacent elements until
 * they are sorted. */
void bubbleSort(int items[], int size) {
	int sorted;
    // Sort items until sorted
    do {
    	sorted = 1; // Assume sorted
        // Swap items consecutively to end of array
        for (int i = size - 1; i > 0; --i) {
            if ( swapAsc( &items[i], &items[i - 1] ) ) {
                sorted = 0;
            }
        }
    } while (!sorted);
}

/* Compares two elements, A and B,
 * and swaps them if they aren't ascending. */
int swapAsc(int *A, int *B) {
    if (*B > *A) {
        int temp = *A;
        *A = *B;
        *B = temp;
        return 1;
    }
    return 0;
}

/** Print to standard out an array of integers.
    Integers are separated by a space. */
void printIntArray(int size, int items[]) {
	for (int i = 0; i < size; i++) {
		printf("%d ", items[i]);
	}
	printf("\n");
}

File: static/algorithms/sorting/bubble_sort/go/bubble_sort.go
package main
import "fmt"

/* bubbleSort Sorts elements by comparing adjacent elements until
 * they are sorted. */
func bubbleSort(items []int) {
    for {
        list_modified := false
        for i := len(items) - 1; i > 0; i-- {
            if items[i] < items[i-1] {
                items[i], items[i-1] = items[i-1], items[i]
                 list_modified = true
            }
        }
        if !list_modified {
            break
        }
    }
}

// Running bubble sort example
func main() {
    items := []int{4, 4, 3, 4, 1}
    fmt.Println("Unsorted items: ", items)
    bubbleSort(items)
    fmt.Println("Sorted items: ", items)
}